File: AWS/iac-microbenchmark/ec2/python/aes256_decrypt.py
#!/usr/bin/env python3
import boto3
import base64
import json
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
import sys

def main():
    # Read JSON input from stdin
    input_data = sys.stdin.read()

# Parse the JSON input
    data = json.loads(input_data)

    # Initialize KMS client
    kms_client = boto3.client('kms')
    
    # Extract the encrypted data key and encrypted message from the data
    encrypted_data_key = base64.b64decode(data.get('encrypted_data_key'))
    ciphertext = base64.b64decode(data.get('encrypted_message'))
    iv = base64.b64decode(data.get('iv'))
    tag = base64.b64decode(data.get('tag'))
    
    # Decrypt the data key using KMS
    response = kms_client.decrypt(
        CiphertextBlob=encrypted_data_key
    )

    plaintext_data_key = response['Plaintext']
    
    # Initialize decryption using AES-GCM
    cipher = Cipher(algorithms.AES(plaintext_data_key), modes.GCM(iv, tag))
    decryptor = cipher.decryptor()
    
    # Decrypt the message
    plaintext_message = decryptor.update(ciphertext) + decryptor.finalize()
    

    decrypt_message = {"Message" : plaintext_message}

    print(json.dumps(decrypt_message))

if __name__ == "__main__":
    main()

File: AWS/iac-microbenchmark/ec2/typescript/aes256_decrypt.ts
#!/usr/bin/env node
import { KMSClient, DecryptCommand } from '@aws-sdk/client-kms';
import * as process from 'process';
import * as crypto from 'crypto';
import * as base64 from 'base64-js';

// Initialize the AWS KMS client with the specified region
const kmsClient = new KMSClient({
  region: 'us-east-1' // Specify the desired region
});

async function main(): Promise<void> {
  try {
    // Get the JSON input from command line arguments
    const requestJsonRaw: string = process.argv[2];

    // Parse the JSON input
    const requestJson = JSON.parse(requestJsonRaw);

    // Extract the encrypted data key, IV, tag, and encrypted message from the parsed JSON
    const encryptedDataKey = base64.toByteArray(requestJson.encrypted_data_key);
    const ciphertext = base64.toByteArray(requestJson.encrypted_message);
    const iv = base64.toByteArray(requestJson.iv);
    const tag = base64.toByteArray(requestJson.tag);

    // Decrypt the data key using KMS
    const decryptCommand = new DecryptCommand({
      CiphertextBlob: encryptedDataKey
    });

    const response = await kmsClient.send(decryptCommand);
    const plaintextDataKey = response.Plaintext as Uint8Array;

    // Initialize decryption using AES-GCM
    const cipher = crypto.createDecipheriv('aes-256-gcm', plaintextDataKey, iv);
    cipher.setAuthTag(tag);

    // Decrypt the message
    const plaintextMessage = Buffer.concat([
      cipher.update(ciphertext),
      cipher.final()
    ]);

    // Prepare the decrypted message for output
    const decryptMessage = {
      Message: plaintextMessage.toString('utf-8') // Convert to string
    };

    // Output the decrypted message as a JSON string
    console.log(JSON.stringify(decryptMessage));
  } catch (error) {
    // Handle and log any errors that occur
    console.error("Error:", error);
  }
}

// Execute the main function
main();
